swift_scality_backend/diskfile.py

```diff
@@ -357,9 +406,38 @@ def create(self, size=None):
         yield DiskFileWriter(self.client_collection, self._name, self.logger)
 
     @utils.trace
-    def write_metadata(self, metadata):
+    def write_metadata(self, md_to_add):
         """Write a block of metadata to an object."""
-        self.client_collection.get_write_client().put_meta(self._name, metadata)
+
+        if not self._metadata:
+            md_to_write = copy.deepcopy(md_to_add)
+            md_to_write.update({
+                'df': {
+                    'name': self._name,
+                },
+                'mf': md_to_add
+            })
+            self.client_collection.get_write_client().put_meta(self._name, md_to_write)
+            return
+
+        # Keep the most recent Content-Type from either datafile or metafile metadata
+        if all(['Content-Type' not in md_to_add,
+                self._metadata is not None,
+                'Content-Type' in self._metadata['mf'],
+                self._metadata['mf']['Content-Type'] != self._metadata['df']['Content-Type'],
+                self._metadata['mf']['Content-Type-Timestamp'] >
+                self._metadata['df']['X-Timestamp']]):
```

Comment:
Restored the initial ugly if after testing: the all() syntax evaluates each of the conditions, without stopping at the first false.

This gives a KeyError on `self._metadata['mf']['Content-Type']` if the previous condition `'Content-Type' in self._metadata['mf']` is wrong